designs/images/cluster-controller-container-components.puml

```diff
@@ -0,0 +1,59 @@
+@startuml cluster-controller-container-components
+!include https://raw.githubusercontent.com/plantuml-stdlib/C4-PlantUML/master/C4_Component.puml
+
+title Component diagram for EKS-A Cluster controller
+
+AddRelTag("indirect relantionship", $textColor="orange", $lineColor="orange", $lineStyle = DashedLine())
+
+Person(user, "EKS-A user", "Manages EKS-A clusters")
+Container(kubeApiServer, "Kubernetes API server", "Go", "Provides read/write access to the kubernetes API objects stored in the etcd server and allows to subscribe to events when they change")
+Container_Ext(workloadKubeApiServer, "Workload Cluster Kubernetes API server", "Go", "Provides read/write access to the kubernetes API objects stored in the etcd server")
+Container_Ext(ociRegistry, "OCI registry", "Public ECR or other", "OCI compliant registry to store container images and Helm charts")
+
+Container_Boundary(clusteController, "Cluster Controller") {
+    AddElementTag("port", $fontColor="#ffffbf", $borderColor="#ffffbf")
+    AddElementTag("busines logic", $fontColor="#01ffff", $borderColor="#01ffff")
+
+    Component(clusterReconciler, "Cluster Reconciler", "K8s controller-runtime Reconciler", "Validates input and orchestrates Cluster reconcile requests", $tags="busines logic")
+    Component(vsphereHandler, "vSphere Cluster Reconciler", "ProviderClusterReconciler", "Reconciles a vSphere Cluster to the desired state", $tags="busines logic")
+    Component(snowHandler, "Snow Cluster Reconciler", "ProviderClusterReconciler", "Reconciles a Snow Cluster to the desired state", $tags="busines logic")
+
+    Component(cilimHandler, "Cilium Reconciler", "", "Reonciles the Cilium CNI installation to the desired state", $tags="busines logic")
+
+    Component(client, "Kubernetes API client", "controller-runtime client.Client", "Kubernetes client provided by controller-runtime", $tags="port")
```

Comment:
I'm thinking we can drop these components. Looking at this now, it feels like it detracts from the core component set (where you'd find the business rules). Instead we can depict a relationship between the components and the KupAPI server (the fact it happens via a client is probably an even lower level diagram).

Reply:
I see what you are saying, however I don't think they detract.

In fact I believe they help understand where the boundaries of the "business rules" components are. These are what decouple the business rules from the protocols and "languages" of the external containers.